engine/crates/composition/src/compose/directives.rs

```diff
@@ -0,0 +1,78 @@
+use super::*;
+
+pub(super) fn collect_composed_directives<'a>(
+    containers: impl Iterator<Item = subgraphs::DirectiveContainerWalker<'a>> + Clone,
+    ctx: &mut ComposeContext<'a>,
+) -> Vec<federated::Directive> {
+    let mut tags: BTreeSet<StringId> = BTreeSet::new();
+    let mut is_inaccessible = false;
+    let mut extra_directives = BTreeSet::new();
+    let mut composed_directives = Vec::new();
+
+    if let Some(deprecated) = containers.clone().find_map(|directives| directives.deprecated()) {
+        composed_directives.push(federated::Directive {
+            name: ctx.insert_static_str("deprecated"),
+            arguments: deprecated
+                .reason()
+                .map(|reason| {
+                    (
+                        ctx.insert_static_str("reason"),
+                        federated::Value::String(ctx.insert_string(reason.id)),
+                    )
+                })
+                .map(|arg| vec![arg])
+                .unwrap_or_default(),
+        });
+    }
+
+    for container in containers {
+        tags.extend(container.tags().map(|t| t.id));
+        is_inaccessible |= container.inaccessible();
```

Comment:
Might be a bit cleaner to do this separately or spell it out more, was a bit surprised to see this

Reply:
yep, I'll see how to make it more readable. At the very least with a comment.

Reply:
done